### DISCUSSION TITLE
rasterize_canopy errors with terra and raster packages

### DISCUSSION THREAD

**Discussion by poornima-s10** [r-lidar/lidR#550] [Q&A]
Hi,

I have recently started using lidR and have run into some issues with rasterize_canopy. I saw that there were several changes in `utils_raster.R` related to functions used in `rasterize_canopy` (functions: `raster_materialize`, `raster_layout` and `raster_set_values`) some months ago. 

I have included some information below, but am unable to share a reproducible example and am looking for some pointers on how to debug this further. 

I have a ground classified/normalised point cloud. I am using lascatalog mainly but in this case have tested both `LAScatalog` and `LAS` objects with `rasterize_canopy` (to rule out any issues with my `LAScatalog` chunks).

This is the command I am using
```r
chm_p2r_ctg = rasterize_canopy(ctg_norm, 0.10, p2r(0.2))
```
(also tested with las object : 
```r
chm_p2r_las = rasterize_canopy(las_norm, 0.10, p2r(0.2)))
```
The error when using the default 'terra' package: 
```
'Error: [] lengths of of cells and values do not match'
```
I tried changing the default to 'raster' using the following command:
```r
options(lidR.raster.default = "raster")
```
This goes through but the output is incorrect, the projection is all wrong:

![image](https://user-images.githubusercontent.com/102007098/159200798-95046b22-7723-486f-b7bc-5fb29f0b00e3.png)

This is the CHM overlaid on an orthomosaic of the site:

![image](https://user-images.githubusercontent.com/102007098/159200844-dcc3c987-4d64-4142-b88d-2b3cb825c73f.png)

The CRS is projected CS GDA2020 MGA Zone 55 (EPSG: 7855). The point spacing is 0.02. Appreciate any advice. Please let me know if more details will be helpful.

Thanks

**Comment by Jean-Romain** [comment]
Please provide a minimal reproducible example. I cannot do anything if I can't repoduce. Thanks

**Reply by poornima-s10** [reply]
Thanks for the reply. I hope this works: https://www.dropbox.com/s/akmhsawjqtvjuqe/clip2_denoised.zip?dl=0
I changed the resolution of the CHM below to recreate the issue on the clip.

```r
las = readLAS(file.path(in_dir, "clip2_denoised.las"))
mycsf <- csf(sloop_smooth=FALSE, cloth_resolution = 0.2, iterations = 500, class_threshold = 0.1)
gnd_las <- classify_ground(las, mycsf)
norm_las <- normalize_height(gnd_las, knnidw(k=10, p=2))

# this is ok in terra/raster - pulse spacing in clip is 0.03 m. In original file it's 0.02 m.
chm = rasterize_canopy(norm_las, res=0.09, p2r(0.2)) 

# recreate issue seen on original file which had pulse spacing 0.02 m with rasterize_canopy res=0.10
# error on terra: Error: [] lengths of of cells and values do not match
chm = rasterize_canopy(norm_las, res=0.06, p2r(0.2)) 

# change to raster
options(lidR.raster.default = "raster")
chm = rasterize_canopy(norm_las, res=0.06, p2r(0.2)) 
# No error with raster but output is incorrect.
plot(chm, col=height.colors(50))
```

**Reply by Jean-Romain** [reply]
Thanks. I reproduced and I opened an issue. Discussion continues in #552
